### ISSUE TITLE
HTTP/2 streams prioritization

### ISSUE THREAD

**Issue by krizhanovsky** [tempesta-tech/tempesta#1196]
According to [RFC 7540 5.3](https://tools.ietf.org/html/rfc7540#section-5.3) we should schedule or service from the cache requests according to their priority and dependency. Reprioritization also must be implemented.

Depends on #309.

**Comment by aleksostapenko** [comment]
Additional discussion regarding possible dependency/priority tree structure is in https://github.com/tempesta-tech/tempesta/pull/1233#issue-265503627.

**Comment by krizhanovsky** [comment]
## TCP send queue and frames prioritization

At the moment Tempesta FW pushes all the HTTP/2 frames to the TCP send queue, just like it does for HTTP/1 pipelining, so a higher priority frame can be pushed to the queue long after many low priority frames in the head of the queue. During prioritization we need to reorder the queue. However, the queue reordering will introduce unnecessary overhead.

### Generic considerations

1. Prioritization should be done in `tcp_write_xmit()` since streams priorities may change while we're processing the TCP send queue
2. Due to skb splitting no any skb can keep data from different streams (HTTP responses), so streams prioritization is equal to skbs prioritization
3. We do our best to minimize the kernel changes, however with QUIC from #724, which we're also going to upstream, streams handling should go to the kernel. With this task we should keep as much logic as possible in Tempesta code base and patch the kernel only in a minimal way, but it'd be good to make the streams foundation in the kernel for QUIC streams.

### Proposal

1. `ss_do_send()` should call a new function instead of `skb_entail()` operating with the TCP write queue. The function should add an skb to a tree (or other data structure depending on the scheduling algorithm) of lists. Each list correponds to a stream (HTTP response).
2. `ss_do_send()` also should call a new function instead of `tcp_push()`, which does seems unnecessary for us tasks (e.g. URG data is never used with HTTP). The function is probably just a wrapper for `__tcp_push_pending_frames()`.
3. `tcp_write_xmit()` should call for HTTP/2 sockets a Tempesta FW prioritization callback instead of `tcp_send_head()`. THe callback will run the scheduling algorithm and return an skb, which it decides to send now. @EvgeniiMekhanik does this callback can be used also in #1394 as the required second callback?

**Comment by krizhanovsky** [comment]
[Comment](https://github.com/tempesta-tech/tempesta/issues/1394#issuecomment-1522967817) from #1394 :

Doesn't block connection and continue to forward Headers frames while flow control limit is exhausted
Doesn't truncate and suspend any frames except DATA ones on reaching close to flow control limit.

Will be done later in the task of stream prioritization.

**Comment by EvgeniiMekhanik** [comment]
Also we should rework `tfw_h2_sk_prepare_xmit` function to `T_FSM_STATE` and make it more clear.
Also we should send all skbs, which are correspond to one frame during one loop iteration in `tcp_write_xmit`

**Comment by EvgeniiMekhanik** [comment]
First of all i think we should rework `tcp_send_head` like this. 
```
tcp_send_head()
{
#ifdef CONFIG_SECURITY_TEMPESTA
if (sock_flag(sk, SOCK_TEMPESTA)) {
/* Use our data structure. */
} else {
/* Use traditional send queue. */
}
#else
#endif
}
```
tcp_send_head is called from several places in TCP code, so we should patch the whole function, not replace it in `tcp_write_xmit`.

Then we should check all places in linux kernel code, where `tcp_send_head` is called and patch them if it is required.

**Comment by EvgeniiMekhanik** [comment]
#How it is done in nginx

**Comment by EvgeniiMekhanik** [comment]
We decide to support RFC 9218, since RFC 7540 is deprecated and there is a big chance that all modern clients will not support it soon. In the RFC 9218 there are two priority parameters `urgency` and `incremental`:
- The urgency (u) parameter is an integer  value between 0 and 7 inclusive, in descending order of priority.
- The incremental (i) parameter value is boolean. It indicates if an HTTP response can be processed incrementally, i.e., provide some meaningful output as chunks of the response arrive. If a client makes concurrent requests with the incremental parameter set to false, there is no benefit in serving responses with the same urgency concurrently because the client is not going to process those responses incrementally.
I think we should implement array (with size 7) of lists of streams. Each array entry contains TWO list head. First list head is a head of list of incremental streams and second is a head of non-incremental streams. New stream with appropriate urgency is added to appropriate list at the end of the list. All requests served in the order of it's urgency, we don't send requests with urgency 1 while there is any data to send for requests with urgency 0. Non incremental requests served one by one, without bandwidth sharing, but we share bandwidth between incremental and non incremental requests.

**Comment by krizhanovsky** [comment]
Could you please check RFC 9218 support for HTTP/2 in `nghttp2`, Chrome and Firefox? With "check" I mean grepping source code and checking the issues, if they open (e.g. https://bugs.chromium.org/p/chromium/issues/detail?id=1404785)

**Comment by EvgeniiMekhanik** [comment]
I check RFC 9218 support in several projects:
- nghttp2 has RFC 9218 support
- Apache has RFC 9218 support
- Chromium and Firefox has RFC 9218 support ONLY for HTTP3. (https://calendar.perfplanet.com/2022/http-3-prioritization-demystified/)

**Comment by EvgeniiMekhanik** [comment]
Closed by 1973
